service/mapper/transitions_test.go

```diff
@@ -0,0 +1,1233 @@
+	})
+
+	t.Run("nominal case no more registers left to index", func(t *testing.T) {
+		t.Parallel()
+
+		index := &mocks.Writer{
+			PayloadsFunc: func(height uint64, paths []ledger.Path, value []*ledger.Payload) error { return nil },
+		}
+
+		tr := &Transitions{
+			log:   testLog,
+			index: index,
+		}
+
+		st := &State{
+			registers: map[ledger.Path]*ledger.Payload{},
+			height:    testHeight,
+			status:    StatusCollected,
+		}
+
+		err := tr.IndexRegisters(st)
+
+		assert.NoError(t, err)
+		assert.Equal(t, StatusIndexed, st.status)
+	})
+
+	t.Run("handles invalid status", func(t *testing.T) {
+		t.Parallel()
+
+		index := &mocks.Writer{
+			PayloadsFunc: func(height uint64, paths []ledger.Path, value []*ledger.Payload) error { return nil },
+		}
+
+		tr := &Transitions{
+			log:   testLog,
+			index: index,
+		}
+
+		st := &State{
+			registers: map[ledger.Path]*ledger.Payload{},
+			height:    testHeight,
+			status:    StatusEmpty,
+		}
+
+		err := tr.IndexRegisters(st)
+
+		assert.Error(t, err)
+	})
+
+	t.Run("handles indexer failure", func(t *testing.T) {
+		t.Parallel()
+
+		index := &mocks.Writer{
+			PayloadsFunc: func(height uint64, paths []ledger.Path, value []*ledger.Payload) error { return mocks.DummyError },
+		}
+
+		tr := &Transitions{
+			log:   testLog,
+			index: index,
+			cfg: Config{
+				IndexPayloads: true,
+			},
+		}
+
+		st := &State{
+			registers: map[ledger.Path]*ledger.Payload{
+				testPath1: testPayload1,
+				testPath2: testPayload2,
+				testPath3: testPayload3,
+				testPath4: testPayload4,
+				testPath5: testPayload5,
+				testPath6: testPayload6,
+			},
+			height: testHeight,
+			status: StatusCollected,
+		}
+
+		err := tr.IndexRegisters(st)
+
+		assert.Error(t, err)
+	})
+
+}
+
+func TestTransitions_ForwardHeight(t *testing.T) {
+	testHeight := uint64(42)
+	testCommit := flow.StateCommitment{0x2a, 0x04, 0x51, 0x3c, 0xc3, 0xc9, 0xa7, 0xf2, 0xec, 0x08, 0x93, 0x56, 0x5f, 0x52, 0xc2, 0x9e, 0x19, 0xf5, 0x58, 0x88, 0x10, 0x11, 0xe1, 0x13, 0x60, 0x43, 0x9e, 0x57, 0x60, 0x18, 0xe3, 0xde}
+
+	t.Run("nominal case", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			firstCalled int
+			lastCalled  int
+		)
+		index := &mocks.Writer{
+			FirstFunc: func(height uint64) error {
+				assert.Equal(t, testHeight, height)
+				firstCalled++
+				return nil
+			},
+			LastFunc: func(height uint64) error {
+				assert.Equal(t, testHeight+uint64(lastCalled), height)
+
+				lastCalled++
+
+				return nil
+			},
+		}
+
+		forest := &mocks.Forest{
+			ResetFunc: func(finalized flow.StateCommitment) {
+				assert.Equal(t, testCommit, finalized)
+			},
+		}
+
+		tr := &Transitions{
+			log:   testLog,
+			index: index,
+			once:  &sync.Once{},
+		}
+
+		st := &State{
+			registers: map[ledger.Path]*ledger.Payload{
+				testPath1: testPayload1,
+				testPath2: testPayload2,
+				testPath3: testPayload3,
+				testPath4: testPayload4,
+				testPath5: testPayload5,
+				testPath6: testPayload6,
+			},
```

Comment:
This should always be empty after indexing. Not sure why you would populate it here. It's only use is to carry register paths/values from collecting to indexing.